src/dns.rs

```diff
@@ -0,0 +1,464 @@
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push(format!(
+                    "DNS server IP {:?} is invalid. Parse Error: {:?}",
+                    s, error
+                )),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(format!(
+                "DNS resolver initialization failed with errors: {:?}",
+                errors
+            ));
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.use_hosts_file = false;
+        options.distrust_nx_responses = false;
+        options.num_concurrent_reqs = 1;
+
+        let (resolver, worker) = AsyncResolver::new(resolve_config, options);
+
+        // spin up worker thread for worker future
+        std::thread::spawn(move || {
+            // Worker futer will finish when all DnsResolvers are dropped.
+            // Then this thread will also end.
+            tokio::runtime::current_thread::run(worker);
+        });
+
+        Ok(DnsResolver(Some(resolver)))
+    }
+
+    /// Resolves host:port address
+    /// Should be only called from outside tokio. From inside tokio, Async version should be used.
+    pub fn resolve_address(&self, s: &str) -> Result<IntoIter<SocketAddr>, io::Error> {
+        // Try to parse as a regular SocketAddr first
+        if let Some(addr) = s.parse().ok() {
+            return Ok(vec![addr].into_iter());
+        }
+
+        // Now 's' should contain name:port
+        let mut parts = s.rsplitn(2, ':');
+        let port = parts
+            .next()
+            .ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("Missing port in: {:?}", s),
+                )
+            })?
+            .parse::<u16>()
+            .map_err(|error| io::Error::new(io::ErrorKind::InvalidInput, error))?;
+        let name = parts
+            .next()
+            .ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("Missing host in: {:?}", s),
+                )
+            })?
+            .parse::<Name>()
+            .map_err(|error| io::Error::new(io::ErrorKind::InvalidInput, error))?;
+
+        let addresses = self.resolve_name(name)?;
```

Comment:
hmm this is blocking I don't think that is the right way to go about it? I imagine we should lazily look up hosts will in a futures task.

Reply:
Yes, it's problematic. There are those sinks who were easily modified for async resolution, and others who weren't, and I didn't want to modify it all at once. Merger with master was already full of conflicts as this pull request touches a lot of places.

But I think it's fine even if they remain blocking, as even before this they were blocking, just on a different resolver.